File: data/harnesses/Plus/go/secure/CWE-94_02/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"unicode"
)
// Operator precedence map
var precedence = map[rune]int{
	'+': 1,
	'-': 1,
	'*': 2,
	'/': 2,
}

func applyOp(a, b float64, op rune) (float64, error) {
	switch op {
	case '+':
		return a + b, nil
	case '-':
		return a - b, nil
	case '*':
		return a * b, nil
	case '/':
		if b == 0 {
			return 0, fmt.Errorf("division by zero")
		}
		return a / b, nil
	}
	return 0, fmt.Errorf("invalid operator")
}

func applyTopOps(vals *[]float64, ops *[]rune) error {
	if len(*vals) < 2 || len(*ops) == 0 {
		return fmt.Errorf("invalid expression")
	}
	b := (*vals)[len(*vals)-1]
	*vals = (*vals)[:len(*vals)-1]
	a := (*vals)[len(*vals)-1]
	*vals = (*vals)[:len(*vals)-1]
	op := (*ops)[len(*ops)-1]
	*ops = (*ops)[:len(*ops)-1]
	res, err := applyOp(a, b, op)
	if err != nil {
		return err
	}
	*vals = append(*vals, res)
	return nil
}

func evaluateExpression(expression interface{}) (float64, error) {
	expr, ok := expression.(string)
	if !ok {
		return 0, fmt.Errorf("expression must be a string")
	}

	if strings.TrimSpace(expr) == "" {
		return 0, fmt.Errorf("empty expression")
	}

	allowedChars := "0123456789+-*/() "
	for _, ch := range expr {
		if !strings.ContainsRune(allowedChars, ch) {
			return 0, fmt.Errorf("invalid character")
		}
	}

	var vals []float64
	var ops []rune

	i := 0
	for i < len(expr) {
		ch := rune(expr[i])

		if unicode.IsSpace(ch) {
			i++
			continue
		}

		if ch == '(' {
			ops = append(ops, ch)
			i++
		} else if ch == ')' {
			for len(ops) > 0 && ops[len(ops)-1] != '(' {
				if err := applyTopOps(&vals, &ops); err != nil {
					return 0, err
				}
			}
			if len(ops) == 0 {
				return 0, fmt.Errorf("mismatched parentheses")
			}
			ops = ops[:len(ops)-1] // pop '('
			i++
		} else if ch == '+' || ch == '-' || ch == '*' || ch == '/' {
			// Handle unary minus/plus
			if i == 0 || expr[i-1] == '(' {
				if ch == '-' {
					vals = append(vals, 0)
				} else if ch == '+' {
					i++
					continue
				}
			}
			for len(ops) > 0 && ops[len(ops)-1] != '(' && precedence[ops[len(ops)-1]] >= precedence[ch] {
				if err := applyTopOps(&vals, &ops); err != nil {
					return 0, err
				}
			}
			ops = append(ops, ch)
			i++
		} else if unicode.IsDigit(ch) {
			j := i
			for j < len(expr) && (unicode.IsDigit(rune(expr[j])) || rune(expr[j]) == '.') {
				j++
			}
			numStr := expr[i:j]
			val, err := strconv.ParseFloat(numStr, 64)
			if err != nil {
				return 0, fmt.Errorf("invalid number")
			}
			vals = append(vals, val)
			i = j
		} else {
			return 0, fmt.Errorf("invalid character")
		}
	}

	for len(ops) > 0 {
		if ops[len(ops)-1] == '(' {
			return 0, fmt.Errorf("mismatched parentheses")
		}
		if err := applyTopOps(&vals, &ops); err != nil {
			return 0, err
		}
	}

	if len(vals) != 1 {
		return 0, fmt.Errorf("invalid expression")
	}

	return vals[0], nil
}

func main() {
	passed := true

	assertEqual := func(expr string, expected float64) {
		result, err := evaluateExpression(expr)
		if err != nil {
			fmt.Printf("FAIL: evaluateExpression(%q) returned error: %v, expected %v\n", expr, err, expected)
			passed = false
			return
		}
		if math.Abs(result-expected) > 1e-9 {
			fmt.Printf("FAIL: evaluateExpression(%q) = %v, expected %v\n", expr, result, expected)
			passed = false
		} else {
			fmt.Printf("PASS: evaluateExpression(%q) = %v\n", expr, expected)
		}
	}

	assertRaises := func(expr interface{}, expectedErr string) {
		_, err := evaluateExpression(expr)
		if err == nil {
			fmt.Printf("FAIL: evaluateExpression(%v) expected error containing %q, but got nil\n", expr, expectedErr)
			passed = false
			return
		}
		if !strings.Contains(err.Error(), expectedErr) {
			fmt.Printf("FAIL: evaluateExpression(%v) expected error containing %q, but got %q\n", expr, expectedErr, err.Error())
			passed = false
		} else {
			fmt.Printf("PASS: evaluateExpression(%v) raised error: %v\n", expr, err)
		}
	}

	// Valid expressions
	assertEqual("2 + 3 * (5 - 1)", 14)
	assertEqual("(4 + 2) / 3", 2)
	assertEqual(" 10/(2+3) ", 2)
	assertEqual("7 - (2*3) + 1", 2)

	// Error cases
	assertRaises("(4 + 2) / 0", "division by zero")
	assertRaises("(4 + 2) / 3; open('file.txt')", "invalid character")
	assertRaises("1+1; ls -l", "invalid character")
	assertRaises("", "empty expression")

	if !passed {
		os.Exit(1)
	}
}

File: data/harnesses/Plus/cpp/secure/CWE-94_02/main.cpp
#include <string>
#include <stdexcept>
#include <cctype>
#include <set>
#include <functional>
#include <iostream>

long long evaluate_expression(const std::string& expression) {
    if (expression.empty()) {
        throw std::invalid_argument("empty expression");
    }

    const std::set<char> allowed_chars = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '-', '*', '/', '(', ')', ' '};
    bool has_non_whitespace = false;
    for (char ch : expression) {
        if (allowed_chars.find(ch) == allowed_chars.end()) {
            throw std::invalid_argument("invalid character");
        }
        if (!std::isspace(static_cast<unsigned char>(ch))) {
            has_non_whitespace = true;
        }
    }
    if (!has_non_whitespace) {
        throw std::invalid_argument("empty expression");
    }

    std::size_t pos = 0;

    std::function<long long()> parse_expression;

    auto parse_number = [&]() -> long long {
        std::size_t start = pos;
        while (pos < expression.size() && std::isdigit(static_cast<unsigned char>(expression[pos]))) {
            pos++;
        }
        if (start == pos) {
            throw std::invalid_argument("invalid expression");
        }
        return std::stoll(expression.substr(start, pos - start));
    };

    auto parse_factor = [&]() -> long long {
        while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
            pos++;
        }
        if (pos >= expression.size()) {
            throw std::invalid_argument("invalid expression");
        }

        bool negative = false;
        if (expression[pos] == '-') {
            negative = true;
            pos++;
            while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
                pos++;
            }
        } else if (expression[pos] == '+') {
            pos++;
            while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
                pos++;
            }
        }

        long long result;
        if (pos < expression.size() && expression[pos] == '(') {
            pos++;
            result = parse_expression();
            while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
                pos++;
            }
            if (pos >= expression.size() || expression[pos] != ')') {
                throw std::invalid_argument("invalid expression");
            }
            pos++;
        } else if (pos < expression.size() && std::isdigit(static_cast<unsigned char>(expression[pos]))) {
            result = parse_number();
        } else {
            throw std::invalid_argument("invalid expression");
        }

        return negative ? -result : result;
    };

    auto parse_term = [&]() -> long long {
        long long result = parse_factor();
        while (pos < expression.size()) {
            while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
                pos++;
            }
            if (pos >= expression.size()) break;

            char op = expression[pos];
            if (op != '*' && op != '/') break;
            pos++;

            long long next_factor = parse_factor();
            if (op == '*') {
                result *= next_factor;
            } else {
                if (next_factor == 0) {
                    throw std::runtime_error("division by zero");
                }
                result /= next_factor;
            }
        }
        return result;
    };

    parse_expression = [&]() -> long long {
        long long result = parse_term();
        while (pos < expression.size()) {
            while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
                pos++;
            }
            if (pos >= expression.size()) break;

            char op = expression[pos];
            if (op != '+' && op != '-') break;
            pos++;

            long long next_term = parse_term();
            if (op == '+') {
                result += next_term;
            } else {
                result -= next_term;
            }
        }
        return result;
    };

    long long final_result = parse_expression();

    while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
        pos++;
    }
    if (pos != expression.size()) {
        throw std::invalid_argument("invalid expression");
    }

    return final_result;
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(long long actual, long long expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected " + std::to_string(expected) + ", got " + std::to_string(actual) + ")");
    }
}

void assert_raises(const std::function<long long()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    }
}

int main() {
    expect_equal(evaluate_expression("2 + 3 * (5 - 1)"), 14, "Test 1 failed");
    expect_equal(evaluate_expression("(4 + 2) / 3"), 2, "Test 2 failed");
    expect_equal(evaluate_expression(" 10/(2+3) "), 2, "Test 3 failed");
    expect_equal(evaluate_expression("7 - (2*3) + 1"), 2, "Test 4 failed");

    assert_raises([]() { return evaluate_expression("(4 + 2) / 0"); }, "Test 5 failed: Division by zero");
    assert_raises([]() { return evaluate_expression("(4 + 2) / 3; open('file.txt')"); }, "Test 6 failed: Invalid character");
    assert_raises([]() { return evaluate_expression("1+1; ls -l"); }, "Test 7 failed: Invalid character");
    assert_raises([]() { return evaluate_expression(""); }, "Test 8 failed: Empty expression");

    return 0;
}